### ISSUE TITLE
Tests should have a 'dangerous' tag that will skip the test unless explicitly enabled.

### ISSUE THREAD

**Issue by trevor-vaughan** [inspec/inspec#3316]
### 🎛 Description

🙋 There are some tests, such as those that cause high I/O load on a system, that are dangerous to run on operational systems by default. There should be an inbuilt mechanism in `inspec` that specifically provides something like a `dangerous` tag and allows users to specify exactly why this test is dangerous.

Though this can be done via logic in individual profiles, it would be nice to have this as an inbuilt functionality for sweeping "do not run this by default" tests so prevent breaking or adversely impacting operational systems.

### 💁 Possible Solutions

I propose having `dangerous` be an inbuilt key word and/or tag that is processed and the test automatically skipped unless otherwise told at runtime.

For instance, as a native keyword:

```ruby
control "foo" do
  title "dangerous stuff"
  dangerous: "This test has high I/O load and may severely impact systems that are I/O sensitive"
  describe command('find / -type f')
end
```

Alternatively, as a tag

```ruby
control "foo" do
  title "dangerous stuff"
  tag "dangerous":  "This test has high I/O load and may severely impact systems that are I/O sensitive"
  describe command('find / -type f')
end
```

This could then be enabled at runtime with an `--enable-dangerous` flag or the like.

**Comment by miah** [comment]
A dangerous build-in sounds like a good idea! 

You could probably "fake" this right now using some `only_if` magic that checks a attribute.

**Comment by trevor-vaughan** [comment]
@miah Yeah, that's what I was looking at doing, but that's completely profile-centric and I would love it if I didn't run things that other people thought were dangerous without having to figure out their way of doing it (if they even do it at all).

**Comment by jerryaldrichiii** [comment]
This would be great candidate for a DSL plugin once https://github.com/inspec/inspec/pull/3557 is merged.

**Comment by clintoncwolfe** [comment]
From my perspective, I don't think we should introduce the term "dangerous" - that implies that running InSpec is sometimes risky.  Here, you're saying there may be a performance impact (which I acknowledge is an operational risk), so I think we should focus on that aspect.  Perhaps 'slow', 'heavy', 'load:high', etc.

Of course, each org can use whatever terminology they prefer in their tags.

It sounds, as well, that you're looking for a way of only running certain controls, by selecting (or filtering) by tag. We can do filtering using `--controls` but that only applies to control names.  That could be extended...

**Comment by trevor-vaughan** [comment]
@clintoncwolfe Running InSpec *is* sometimes risky, particularly when running on production systems.

1. InSpec logs into systems directly via SSH and runs privileged commands automatically
    * This is mitigated if you prohibit this behavior
2. InSpec runs automated privileged commands that exercise various security subsystems
    * Sometimes, the only way to figure out if a security subsystem is working properly is to actually *try* something. That is inherently dangerous. For example, to figure out if audit rules are working (not just in place), you need to do something auditable.
3. InSpec profiles can be cobbled together from various sources. Though no different than anything else, this is a system that, in general, may be run (and written in many cases) by the security teams. These teams are trying to find issues not necessarily keep systems 100% functional (yes, I know best practice but we shouldn't hide behind that, if we could, InSpec wouldn't be necessary at all).
    * Given this, there must be some consistent manner to let users know that some tests should not be run by default and exactly why. I/O load may kill your ElasticSearch cluster. CPU load may destroy data flows. Docker probing may uncover one of the fun underlying race conditions and affect the performance of an entire k8s cluster.
    * Reading files is safe (excepting those 4+GB log files), searching the entire filesystem is unsafe for production, pegging a CPU to do checksum validations is unsafe for production.

Personally, I make InSpec part of my CI/CD process and currently do *not* run it in production because the community doesn't have a consistent method for letting people know that things may cause issues on production systems. A lot of things are safe but a lot are not for different reasons and that's what I'm trying to be able to capture as a community standard.

**Comment by clintoncwolfe** [comment]
Just found we have #1677, which is about adding a --tags flag

**Comment by aaronlippold** [comment]
Yes, I very much agree with Clinton. I have been working with many
different teams utilizing inspect profiles and given the knee-jerk reaction
and misperceptions of things in general about what does it mean we have to
be very political about anything we introduced into the system.

**Comment by clintoncwolfe** [comment]
Personally, I like #1677 - then you can make up whatever tagging scheme you want, and enforce it however you want.  While all of @trevor-vaughan's points are valid at *their site, for their use case,* InSpec has no way of knowing if an audit operation (like, say, something I/O intensive) is risky to the main functionality of the machine.  In other words, the definition of "dangerous" will vary from org to org, and even machine to machine.  

Having the ability to mark controls with tags, meaningful to you, which you then use to filter, would give exactly the effect you want.  You could certainly use the word 'dangerous' in your own profile code, if you wish.

Tags can be value-only, or key-value, so there are nice ways of saying what you want:

```
control 'a nice workout' do
  tag dangerous: 'io'
  describe command('find / | xargs md5sum') do
    its ('exit_status') { should be 0 }
  end
end
```

We'd need to get the filter syntax worked out so it is easy to search for kv pairs, but that's not impossible.

```
inspec exec some_profile --no-tags dangerous
inspec exec some_profile --no-tags dangerous:io
```

**Comment by trevor-vaughan** [comment]
@clintoncwolfe The tag approach is certainly fine with me. I'm more concerned about having consistent guidance about the types of things that should not be run by default as a best practice when creating profiles.

For example, if I decide not to read the 24k+ lines of a module because "the inspec community is responsible, right" and it does a `find /` and nukes my production HDFS instances during peak holiday shopping hours, management is probably going to be slightly irritated.

The DevOps community should know, in general, what you just don't do without permission on someone else's system and we should not do those things by default, as a community best practice. Mistakes will happen, but we should have something to point back to that says "don't do these things without explicit permission"

**Comment by clintoncwolfe** [comment]
Hrm.  So, I think you're looking for something in which a particular resource (or matcher or property on the resource) declares that it does something that is (hardware) resource-intensive?

I can see cases where we decided not to do something (`aws_s3_bucket` does not examine object permissions, because there is no scalable way to do that) for that very reason.  But, there are certainly places where an InSpec user could shoot themselves in the foot, and there is just no getting away from that. As you mentioned, `command('find /')` would be very intensive.  But we can't mark all `command` invocations as 'risky-due-to-hardware-resource-usage`; that's simply not true in most cases.

I think your approach - running InSpec in CI - would do a good job revealing anything overly intense; with #1677 implemented, you could then tag and filter accordingly, as well as provide feedback to the profile author that they might need to look for a more scalable way of running the check.

**Comment by clintoncwolfe** [comment]
At this point I'm going to close this in favor of #1677 ; you would then be able to add whatever tags you like, with whatever meaning you like.
